core/src/main/java/org/pentaho/di/core/plugins/PluginRegistry.java

```diff
@@ -0,0 +1,1114 @@
+        return (T) ( (ClassLoadingPluginInterface) plugin ).getClassLoader().loadClass( className );
+      } else {
+        URLClassLoader ucl;
+        lock.writeLock().lock();
+        try {
+          Map<PluginInterface, URLClassLoader> classLoaders =
+            classLoaderMap.computeIfAbsent( plugin.getPluginType(), k -> new HashMap<>() );
+          ucl = classLoaders.get( plugin );
+
+          if ( ucl == null ) {
+            if ( !Utils.isEmpty( plugin.getClassLoaderGroup() ) ) {
+              ucl = classLoaderGroupsMap.get( plugin.getClassLoaderGroup() );
+            } else {
+              ucl = folderBasedClassLoaderMap.get( plugin.getPluginDirectory().toString() );
+            }
+          }
+          if ( ucl != null ) {
+            classLoaders.put( plugin, ucl ); // save for later use...
+          }
+        } finally {
+          lock.writeLock().unlock();
+        }
+
+        if ( ucl == null ) {
+          throw new KettlePluginException( "Unable to find class loader for plugin: " + plugin );
+        }
+        return (T) ucl.loadClass( className );
+
+      }
+    } catch ( Exception e ) {
+      throw new KettlePluginException( "Unexpected error loading class with name: " + className, e );
+    }
+  }
+
+  /**
+   * Load the class with a certain name using the class loader of certain plugin.
+   *
+   * @param plugin    The plugin for which we want to use the class loader
+   * @param classType The type of class to load
+   * @return the name of the class
+   * @throws KettlePluginException In case there is something wrong
+   */
+  @SuppressWarnings( "unchecked" )
+  public <T> T getClass( PluginInterface plugin, T classType ) throws KettlePluginException {
+    String className = plugin.getClassMap().get( classType );
+    return (T) getClass( plugin, className );
+  }
+
+  /**
+   * Create or retrieve the class loader for the specified plugin
+   *
+   * @param plugin the plugin to use
+   * @return The class loader
+   * @throws KettlePluginException In case there was a problem
+   *                               <p/>
+   *                               getClassLoader();
+   */
+  public ClassLoader getClassLoader( PluginInterface plugin ) throws KettlePluginException {
+
+    if ( plugin == null ) {
+      throw new KettlePluginException( BaseMessages.getString(
+          PKG, "PluginRegistry.RuntimeError.NoValidStepOrPlugin.PLUGINREGISTRY001" ) );
+    }
+
+    try {
+      if ( plugin.isNativePlugin() ) {
+        return this.getClass().getClassLoader();
+      } else {
+        URLClassLoader ucl;
+
+        lock.writeLock().lock();
+        try {
+          // If the plugin needs to have a separate class loader for each instance of the plugin.
+          // This is not the default. By default we cache the class loader for each plugin ID.
+          if ( plugin.isSeparateClassLoaderNeeded() ) {
+            // Create a new one each time
+            ucl = createClassLoader( plugin );
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
Analysis of this package determined that this line of code might contain a resource leak. A resource leak can slow down or crash your system.

**Fix**
Consider closing the following resource in a try-finally block: *ucl*. Execution paths that do not contain closure statements were detected.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).